src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -426,10 +427,21 @@ def get_installation_error(self, e):
                 req_disp = str(req)
             else:
                 req_disp = f'{req} (from {parent.name})'
+
+            cands = self._finder.find_all_candidates(req.project_name)
```

Comment:
This is potentially quite costly. Of course, the finder caches results from `find_all_candidates`, and presumably we called it earlier for this project, but is it OK to rely on that caching? I could imagine someone profiling the resolver, and coming to the conclusion that the cache isn't helping, and removing it without noticing the impact here.

At a minimum, this feels like it's relying on an implementation detail (that the resolver calls `find_all_candidates` via `find_best_candidate`), so even if we do it for now, at some point it should be refactored to make the dependency explicit.

Reply:
@pfmoore I had the exact same thought earlier, but in the current state of things it was difficult to be explicit about it.

The reason I left it like this was that `finder.find_best_candidate` (used in resolvelib) uses `find_all_candidates` under the hood, so without refactoring this relationship out, there's no way of removing the `lru_cache` without making resolver dead slow.

Reply:
... which is basically why we dropped the reporting (I'm not sure if we had the LRU cache when we first wrote that code). Exploiting the cache is a neat way of getting the reporting back, and I'm fine with it up to a point, but I have concerns about maintainability as it's a hidden, and subtle, dependency.

I have no better suggestion that isn't a major refactor, though, so I agree it's "good enough for now".